File: aws_three_tier_OTA_project/frontend/src/services/authService.ts
import { apiClient, handleApiError } from './api';
import type {
  LoginRequest,
  RegisterRequest,
  AuthResponse,
  User,
  Traveler,
  ApiResponse,
} from '@/types';

export const authService = {
  async login(credentials: LoginRequest): Promise<AuthResponse> {
    try {
      const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/login', credentials);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async register(data: RegisterRequest): Promise<AuthResponse> {
    try {
      const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/register', data);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async logout(): Promise<void> {
    try {
      await apiClient.post('/auth/logout');
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async refreshToken(refreshToken: string): Promise<AuthResponse> {
    try {
      const response = await apiClient.post<ApiResponse<AuthResponse>>('/auth/refresh', {
        refreshToken,
      });
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async getCurrentUser(): Promise<User> {
    try {
      const response = await apiClient.get<ApiResponse<User>>('/auth/me');
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async updateProfile(data: Partial<User>): Promise<User> {
    try {
      const response = await apiClient.put<ApiResponse<User>>('/auth/profile', data);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async changePassword(currentPassword: string, newPassword: string): Promise<void> {
    try {
      await apiClient.post('/auth/change-password', {
        currentPassword,
        newPassword,
      });
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async getSavedTravelers(): Promise<Traveler[]> {
    try {
      const response = await apiClient.get<ApiResponse<Traveler[]>>('/travelers');
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async addTraveler(traveler: Omit<Traveler, 'id' | 'userId'>): Promise<Traveler> {
    try {
      const response = await apiClient.post<ApiResponse<Traveler>>('/travelers', traveler);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async updateTraveler(id: string, traveler: Partial<Traveler>): Promise<Traveler> {
    try {
      const response = await apiClient.put<ApiResponse<Traveler>>(`/travelers/${id}`, traveler);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  async deleteTraveler(id: string): Promise<void> {
    try {
      await apiClient.delete(`/travelers/${id}`);
    } catch (error) {
      throw handleApiError(error);
    }
  },
};

File: aws_three_tier_OTA_project/backend/auth-service/src/main/java/com/travelease/auth/service/AuthService.java
package com.travelease.auth.service;

import com.travelease.auth.dto.*;
import com.travelease.auth.entity.User;
import com.travelease.auth.repository.UserRepository;
import com.travelease.common.exception.BusinessException;
import com.travelease.common.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        log.info("Registering new user with email: {}", request.getEmail());

        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BusinessException("Email already registered");
        }

        User user = User.builder()
                .email(request.getEmail().toLowerCase())
                .password(passwordEncoder.encode(request.getPassword()))
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .phone(request.getPhone())
                .role(User.Role.USER)
                .enabled(true)
                .emailVerified(false)
                .phoneVerified(false)
                .build();

        user = userRepository.save(user);
        log.info("User registered successfully with ID: {}", user.getId());

        return generateAuthResponse(user);
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {
        log.info("User login attempt for email: {}", request.getEmail());

        User user = userRepository.findByEmail(request.getEmail().toLowerCase())
                .orElseThrow(() -> new BusinessException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new BusinessException("Invalid email or password");
        }

        if (!user.getEnabled()) {
            throw new BusinessException("Account is disabled");
        }

        userRepository.updateLastLoginAt(user.getId(), LocalDateTime.now());
        user.setLastLoginAt(LocalDateTime.now());

        log.info("User logged in successfully: {}", user.getId());
        return generateAuthResponse(user);
    }

    @Transactional
    public AuthResponse refreshToken(String refreshToken) {
        if (!jwtTokenProvider.validateToken(refreshToken)) {
            throw new BusinessException("Invalid or expired refresh token");
        }

        String userId = jwtTokenProvider.getUserIdFromToken(refreshToken);
        User user = userRepository.findById(UUID.fromString(userId))
                .orElseThrow(() -> new BusinessException("User not found"));

        if (!user.getEnabled()) {
            throw new BusinessException("Account is disabled");
        }

        log.info("Token refreshed for user: {}", user.getId());
        return generateAuthResponse(user);
    }

    @Transactional(readOnly = true)
    public UserResponse getCurrentUser(String userId) {
        User user = userRepository.findById(UUID.fromString(userId))
                .orElseThrow(() -> new BusinessException("User not found"));

        return UserResponse.fromEntity(user);
    }

    @Transactional
    public UserResponse updateProfile(String userId, UpdateProfileRequest request) {
        User user = userRepository.findById(UUID.fromString(userId))
                .orElseThrow(() -> new BusinessException("User not found"));

        if (request.getFirstName() != null) {
            user.setFirstName(request.getFirstName());
        }
        if (request.getLastName() != null) {
            user.setLastName(request.getLastName());
        }
        if (request.getPhone() != null) {
            user.setPhone(request.getPhone());
        }
        if (request.getProfilePictureUrl() != null) {
            user.setProfilePictureUrl(request.getProfilePictureUrl());
        }
        if (request.getPreferences() != null) {
            user.setPreferences(request.getPreferences());
        }

        user = userRepository.save(user);
        log.info("Profile updated for user: {}", user.getId());

        return UserResponse.fromEntity(user);
    }

    @Transactional
    public void changePassword(String userId, ChangePasswordRequest request) {
        User user = userRepository.findById(UUID.fromString(userId))
                .orElseThrow(() -> new BusinessException("User not found"));

        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPassword())) {
            throw new BusinessException("Current password is incorrect");
        }

        userRepository.updatePassword(user.getId(), passwordEncoder.encode(request.getNewPassword()));
        log.info("Password changed for user: {}", user.getId());
    }

    private AuthResponse generateAuthResponse(User user) {
        String accessToken = jwtTokenProvider.generateToken(user.getId().toString(), user.getEmail());
        String refreshToken = jwtTokenProvider.generateRefreshToken(user.getId().toString());

        return AuthResponse.of(
                accessToken,
                refreshToken,
                jwtTokenProvider.getAccessTokenExpirationMs(),
                UserResponse.fromEntity(user)
        );
    }
}
